src/NumericalAlgorithms/Spectral/SwshInterpolation.cpp

```diff
@@ -0,0 +1,557 @@
+            return ClenshawRecurrenceConstants<Spin>{local_l_max};
+          });
+  return lazy_clenshaw_cache(l_max);
+}
+
+SwshInterpolator::SwshInterpolator(const DataVector& theta,
+                                   const DataVector& phi,
+                                   const size_t l_max) noexcept
+    : l_max_{l_max},
+      raw_libsharp_coefficient_buffer_{
+          size_of_libsharp_coefficient_vector(l_max)},
+      raw_goldberg_coefficient_buffer_{square(l_max + 1)} {
+  cos_m_phi_ = std::vector<DataVector>(l_max + 1);
+  sin_m_phi_ = std::vector<DataVector>(l_max + 1);
+  cos_theta_ = cos(theta);
+  sin_theta_ = sin(theta);
+  cos_theta_over_two_ = cos(theta / 2.0);
+  sin_theta_over_two_ = sin(theta / 2.0);
+  // evaluate cos(m phi) and sin(m phi) via recurrence
+  cos_m_phi_[0] = DataVector{phi.size(), 1.0};
+  sin_m_phi_[0] = DataVector{phi.size(), 0.0};
+  const DataVector m_phi_beta = sin(phi);
+  const DataVector m_phi_alpha = 2.0 * square(sin(0.5 * phi));
+  for (size_t m = 1; m <= l_max; ++m) {
+    cos_m_phi_[m] = cos_m_phi_[m - 1] - (m_phi_alpha * cos_m_phi_[m - 1] +
+                                         m_phi_beta * sin_m_phi_[m - 1]);
+    sin_m_phi_[m] = sin_m_phi_[m - 1] - (m_phi_alpha * sin_m_phi_[m - 1] -
+                                         m_phi_beta * cos_m_phi_[m - 1]);
+  }
+}
+
+template <int Spin>
+void SwshInterpolator::interpolate(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, Spin>*> interpolated,
+    const SpinWeighted<ComplexModalVector, Spin>& goldberg_modes) const
+    noexcept {
+  interpolated->destructive_resize(cos_theta_.size());
+  interpolated->data() = 0.0;
+
+  // used only if s=0;
+  SpinWeighted<ComplexDataVector, Spin> cached_base_harmonic;
+
+  // used during both recurrence legs
+  SpinWeighted<ComplexDataVector, Spin> current_cached_harmonic;
+  SpinWeighted<ComplexDataVector, Spin> current_cached_harmonic_l_plus_one;
+
+  // perform the Clenshaw sums over positive m >= 0.
+  for (int m = 0; m <= static_cast<int>(l_max_); ++m) {
+    if (abs(Spin) >= abs(m)) {
+      direct_evaluation_swsh_at_l_min(make_not_null(&current_cached_harmonic),
+                                      m);
+      evaluate_swsh_at_l_min_plus_one(
+          make_not_null(&current_cached_harmonic_l_plus_one),
+          current_cached_harmonic, m);
+    } else {
+      evaluate_swsh_m_recurrence_at_l_min(
+          make_not_null(&current_cached_harmonic), m);
+      evaluate_swsh_at_l_min_plus_one(
+          make_not_null(&current_cached_harmonic_l_plus_one),
+          current_cached_harmonic, m);
+    }
+    if (Spin == 0 and m == 0) {
+      cached_base_harmonic = current_cached_harmonic;
+    }
+    clenshaw_sum(interpolated, current_cached_harmonic,
+                 current_cached_harmonic_l_plus_one, goldberg_modes, m);
+  }
+  // perform the Clenshaw sums over m < 0.
+  for (int m = -1; m >= -static_cast<int>(l_max_); --m) {
+    if (m == -1 and Spin == 0) {
```

Comment:
Yes, but that overwrite requires this base case to work, because the pointer return is also used as an input for the recurrence function.

Reply:
To make sure I understand, this is needed to initialize the recurrence? If so, it would be good to add a C++ comment so someone in the future will be able to figure that out more easily :)